build.rs

```diff
@@ -11,6 +17,148 @@ compile_error!("One of the features `pio` or `native` must be selected.");
 #[cfg_attr(all(feature = "pio", not(feature = "native")), path = "build_pio.rs")]
 mod build_impl;
 
+pub(crate) const STABLE_PATCHES: &[&str] = &[
+    "patches/missing_xtensa_atomics_fix.diff",
+    "patches/pthread_destructor_fix.diff",
+    "patches/ping_setsockopt_fix.diff",
+];
+
+#[allow(unused)]
+pub(crate) const MASTER_PATCHES: &[&str] = &[
+    "patches/master_missing_xtensa_atomics_fix.diff",
+];
+
+pub(crate) struct EspIdfBuildOutput<I>
+where
+    I: Iterator<Item = (String, kconfig::Value)>,
+{
+    pub(crate) cincl_args: build::CInclArgs,
+    pub(crate) link_args: Option<build::LinkArgs>,
+    pub(crate) kconfig_args: I,
+    pub(crate) bindgen: bindgen::Factory,
+}
+
+struct EspIdfVersion {
+    major: u32,
+    minor: u32,
+    patch: u32,
+}
+
+impl EspIdfVersion {
+    fn parse(bindings_file: impl AsRef<Path>) -> Result<Self> {
+        let bindings_content = fs::read_to_string(bindings_file.as_ref())?;
+
+        Ok(Self {
+            major: Self::grab_const(&bindings_content, "ESP_IDF_VERSION_MAJOR", "u32")?,
+            minor: Self::grab_const(&bindings_content, "ESP_IDF_VERSION_MINOR", "u32")?,
+            patch: Self::grab_const(bindings_content, "ESP_IDF_VERSION_PATCH", "u32")?,
+        })
+    }
+
+    fn get_cfg(&self) -> impl Iterator<Item = String> {
+        once(format!(
+            "esp_idf_full_version=\"{}.{}.{}\"",
+            self.major, self.minor, self.patch
+        ))
+        .chain(once(format!(
+            "esp_idf_version=\"{}.{}\"",
+            self.major, self.minor
+        )))
+        .chain(once(format!("esp_idf_major_version=\"{}\"", self.major)))
+        .chain(once(format!("esp_idf_minor_version=\"{}\"", self.minor)))
+        .chain(once(format!("esp_idf_patch_version=\"{}\"", self.patch)))
+    }
+
+    fn grab_const<T>(text: impl AsRef<str>, const_name: impl AsRef<str>, const_type: impl AsRef<str>) -> Result<T> where T: FromStr, T::Err: error::Error + Send + Sync + 'static {
+        let const_name = const_name.as_ref();
+
+        let value = regex::Regex::new(&format!(r"\s+const\s+{}\s*:\s*{}\s*=\s*(\S+)\s*;", const_name, const_type.as_ref()))?
+            .captures(text.as_ref())
+            .ok_or_else(|| anyhow!("Failed to capture constant {}", const_name))?
+            .get(1)
+            .ok_or_else(|| anyhow!("Failed to capture the value of constant {}", const_name))?
+            .as_str()
+            .parse::<T>()?;
+
+        Ok(value)
```

Comment:
I think a better approach would be to use [`bindgen::callbacks::ParseCallbacks`](https://docs.rs/bindgen/0.59.1/bindgen/callbacks/trait.ParseCallbacks.html) directly. There we can just compare every string macro for the ones that we need, this should be much faster and doesn't rely on the generated output.

But lets do that in a follow-up commit.

Reply:
Maybe. It is not (just) about string macros of course, as the `#ESP_VERSION_` ones are - in fact - int macros, but `ParseCallbacks` have handlers for both (but not for float macros; strange).

In any case that would be hijacking a bit the original purpose of these callbacks which is essentially Rust type control - for our own purposes. Not very clean but probably cleaner than regexp grepping indeed.

As you say - definitely future, but I'll put a TODO item on top of the `grab_constant` function.